app/src/main/java/org/jellyfin/mobile/car/MediaService.kt

```diff
@@ -0,0 +1,609 @@
+        }
+    }
+
+    override fun onDestroy() {
+        mediaSession.run {
+            isActive = false
+            release()
+        }
+
+        exoPlayer.release()
+    }
+
+    override fun onGetRoot(
+        clientPackageName: String,
+        clientUid: Int,
+        rootHints: Bundle?
+    ): BrowserRoot? {
+        val rootExtras = Bundle().apply {
+            putBoolean(CONTENT_STYLE_SUPPORTED, true)
+            putInt(CONTENT_STYLE_BROWSABLE_HINT, CONTENT_STYLE_LIST_ITEM_HINT_VALUE)
+            putInt(CONTENT_STYLE_PLAYABLE_HINT, CONTENT_STYLE_LIST_ITEM_HINT_VALUE)
+        }
+
+        return BrowserRoot(MediaItemType.Libraries.toString(), rootExtras)
+    }
+
+    /**
+     * The parent id will be in various formats such as
+     * Libraries
+     * Library|{id}
+     * LibraryAlbums|{id}
+     * Album|{id}
+     */
+    override fun onLoadChildren(parentId: String, result: Result<MutableList<MediaItem>>) {
+        if (appPreferences.instanceUrl == null || appPreferences.instanceAccessToken == null || appPreferences.instanceUserId == null) {
+            // the required properties for calling the api client are not available
+            // this should only occur if the user does not have a server connection in the app
+            result.sendError(null)
+            return
+        }
+
+        // ensure the api client is up to date
+        apiClient.ChangeServerLocation(appPreferences.instanceUrl)
+        apiClient.SetAuthenticationInfo(
+            appPreferences.instanceAccessToken,
+            appPreferences.instanceUserId
+        )
+
+        when (parentId) {
+            /**
+             * View that shows all the music libraries
+             */
+            MediaItemType.Libraries.toString() -> {
+                apiClient.GetUserViews(
+                    appPreferences.instanceUserId,
+                    object : Response<ItemsResult>() {
+                        override fun onResponse(response: ItemsResult?) {
+                            if (response == null) {
+                                result.sendError(null)
+                                return
+                            }
+
+                            val views = response.items
+                                .filter { item -> item.collectionType == "music" }
+                                .map { item ->
+                                    val itemImageUrl =
+                                        "%s/Items/%s/Images/Primary?maxWidth=1080&quality=90"
+                                            .format(
+                                                appPreferences.instanceUrl, item.id
+                                            )
+
+                                    val description: MediaDescriptionCompat =
+                                        MediaDescriptionCompat.Builder().apply {
+                                            setMediaId(MediaItemType.Library.toString() + "|" + item.id)
```

Comment:
Only pass the item id as media id, if you need the library id you can pass that in a bundle using `setExtras`

Reply:
(Same for the other places that used this way of communicating data)

Reply:
Unfortunately I don't think this could work. The `onLoadChildren` method only has that `parentId` string arg. This means the media id has to have all the information required to load the next view, which is why I'm playing with strings.

There is an overload method of [onLoadChildren with a bundle](https://developer.android.com/reference/androidx/media/MediaBrowserServiceCompat#onLoadChildren(java.lang.String,%20androidx.media.MediaBrowserServiceCompat.Result%3Cjava.util.List%3Candroid.support.v4.media.MediaBrowserCompat.MediaItem%3E%3E,%20android.os.Bundle)), but I never saw it called while testing. There's a [stackoverflow ](https://stackoverflow.com/questions/55863290/in-mediabrowserservice-onloadchildren-when-how-is-the-options-bundle-supplied)post about it but it has no answer.